File: src/ServerScriptService/Players.server.lua
----------------------
-- Name: Players
-- Authors: MianReplicate
-- Created: 12/31/2024
-- Purpose: To handle players
----------------------
local Modules = script.Parent:WaitForChild("Modules")
local GameLibrary = require(Modules:WaitForChild("GameLibrary"))
local DataCreator = require(Modules:WaitForChild("DataCreator"))
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Settings = ReplicatedStorage:WaitForChild("Settings")
local RModules = ReplicatedStorage:WaitForChild("Modules")
local Util = require(RModules:WaitForChild("Util"))
local Players = game:GetService("Players")

local commands = {}

function commands.print(player : Player, ...)
	return ...
end

function commands.becomerole(player : Player, role : string, optionalPlayerName : string)
	player = (optionalPlayerName and Util.getPlayerFromName(optionalPlayerName, true)) or player
	GameLibrary.changeIntoRole(player, role)
end

function commands.createTool(player : Player, tool : string)
	
end

local prefix = "/"

local function handleCommand(player : Player, message : string)
	if(not Util.isAdmin(player)) then return end
	
	local indexStart = message:find(prefix)
	if(indexStart == 1) then
		message = message:sub(2)
		local split = message:split(" ")
		local command = split[1]:lower()
		table.remove(split, 1)
		
		local commandFunc : (...any) -> any = commands[command]
		if(commandFunc) then
			commandFunc(player, table.unpack(split))
		else
			warn(`{command} is not a valid command! | Used by {player}`)
		end
	end
end

local function saveData(playerData : Data.Data)
	
end

Players.PlayerAdded:Connect(function(player)
	--local playerData : Data.Data = DataCreator.new("Players", player)
	--local function isOwner(_player : Player) return _player == player end
	--playerData:setValue("Coins", 0)
	--playerData:setKeyChecks({"Coins"}, isOwner)
	--playerData:setAccessCheck(isOwner)
	if(Settings:GetAttribute("Testing")) then
		if(not Util.isTester(player)) then player:Kick("Must be a tester to join!") end
	end
	
	player.Chatted:Connect(function(message: string)
		handleCommand(player, message)
	end)
end)

Players.PlayerRemoving:Connect(function(player)
	
end)

File: src/server/Players.server.ts
import { Players, ReplicatedStorage } from "@rbxts/services";
import { Util } from "shared/Util";
import { ActorManager } from "./Managers/ActorManager";
import { TickManager } from "shared/Managers/TickManager";
import { ActorType } from "shared/Types";

type Command = (player: Player, args: string[]) => unknown;

const Settings = ReplicatedStorage.WaitForChild("Settings");
const prefix = "/";
const commands = new Map<string, Command>();

commands.set("becomerole", (player, args) => {
	const role = args[0].lower();
	const optionalPlayerName = args[1];
	player = (optionalPlayerName && Util.getPlayerFromName(optionalPlayerName)) || player;

	ActorManager.changeIntoRole("survivor".match(role)[0] ? ActorType.Survivor : ActorType.Survivor, player);
});

commands.set("tick", (player, args) => {
	const start = args[0].lower();
	if (start === "true") {
		TickManager.setTicking(true);
	} else if (start === "false") {
		TickManager.setTicking(false);
	}
});

function handleCommand(player: Player, message: string) {
	if (!Util.isAdmin(player)) return;

	const indexStart = message.find(prefix)[0];
	if (indexStart === 1) {
		message = message.sub(2);
		const split = message.split(" ");
		const command = split[0].lower();
		split.remove(0);

		const commandFunc = commands.get(command);
		if (commandFunc !== undefined) {
			try {
				commandFunc(player, split);
			} catch (error) {
				print("An error occurred when running the", command, "command!", error);
			}
		} else {
			warn(command + ` is not a valid command! | Used by ` + player.Name);
		}
	}
}

Players.PlayerAdded.Connect((player) => {
	if (Settings.GetAttribute("Testing")) {
		if (!Util.isTester(player)) player.Kick("Must be a tester to join!");
	}

	player.Chatted.Connect((message) => handleCommand(player, message));
	player.CharacterAdded.Connect((character) =>
		character
			.GetDescendants()
			.filter((value) => value.IsA("BasePart"))
			.forEach((value) => (value.CollisionGroup = "Characters")),
	);
});
